Userland/Libraries/LibWeb/SVG/SVGUseElement.cpp

```diff
@@ -0,0 +1,135 @@
+#include <LibWeb/DOM/ElementFactory.h>
+#include <LibWeb/DOM/Event.h>
+#include <LibWeb/DOM/ShadowRoot.h>
+#include <LibWeb/Namespace.h>
+#include <LibWeb/SVG/AttributeNames.h>
+#include <LibWeb/SVG/SVGSVGElement.h>
+#include <LibWeb/SVG/SVGUseElement.h>
+
+namespace Web::SVG {
+
+SVGUseElement::SVGUseElement(DOM::Document& document, DOM::QualifiedName qualified_name)
+    : SVGGraphicsElement(document, qualified_name)
+{
+}
+
+JS::ThrowCompletionOr<void> SVGUseElement::initialize(JS::Realm& realm)
+{
+    MUST_OR_THROW_OOM(Base::initialize(realm));
+    set_prototype(&Bindings::ensure_web_prototype<Bindings::SVGUseElementPrototype>(realm, "SVGUseElement"));
+
+    m_document_observer = TRY(realm.heap().allocate<DOM::DocumentObserver>(realm, realm, document()));
+    m_document_observer->document_fully_loaded = [this]() {
+        create_shadow_tree_of_referenced_element();
+    };
+
+    return {};
+}
+
+void SVGUseElement::visit_edges(Cell::Visitor& visitor)
+{
+    Base::visit_edges(visitor);
+    visitor.visit(m_document_observer);
+}
+
+void SVGUseElement::parse_attribute(DeprecatedFlyString const& name, DeprecatedString const& value)
+{
+    Base::parse_attribute(name, value);
+
+    // https://svgwg.org/svg2-draft/struct.html#UseLayout
+    if (name == SVG::AttributeNames::x) {
+        m_x = AttributeParser::parse_coordinate(value);
+    } else if (name == SVG::AttributeNames::y) {
+        m_y = AttributeParser::parse_coordinate(value);
+    }
+}
+
+void SVGUseElement::inserted()
+{
+    Base::inserted();
+
+    // The x and y properties define an additional transformation (translate(x,y), where x and y represent the computed value of the corresponding property)
+    // to be applied to the ‘use’ element, after any transformations specified with other properties
+    m_transform.translate(m_x.value_or(0), m_y.value_or(0));
+}
+
+// https://svgwg.org/svg2-draft/struct.html#UseShadowTree
+void SVGUseElement::create_shadow_tree_of_referenced_element()
+{
+    // The shadow tree is open (inspectable by script), but read-only.
+    auto shadow_root = heap().allocate<DOM::ShadowRoot>(realm(), document(), *this, Bindings::ShadowRootMode::Open).release_allocated_value_but_fixme_should_propagate_errors();
+
+    // The user agent must create a use-element shadow tree whose host is the ‘use’ element itself
+    set_shadow_root(shadow_root);
+
+    // FIXME: Support the xlink:href attribute as a fallback
+    auto href = attribute(SVG::AttributeNames::href);
+
+    // FIXME: Support loading of external svg documents
+    auto referenced_element = document().query_selector(href.view()).release_value_but_fixme_should_propagate_errors();
+    if (!referenced_element || !is_valid_reference_element(referenced_element)) {
+        return;
+    }
+
+    // The ‘use’ element references another element, a copy of which is rendered in place of the ‘use’ in the document.
+    auto cloned_reference_node = referenced_element->clone_node(nullptr, true);
```

Comment:
If the referenced element subtree is mutated with JavaScript, aren't those mutations supposed to be visible in all the `<use>` instances?

Reply:
Implemented as discussed on discord.

`SVGElement::inserted()` etc. now call into any `SVGUseElement`s and the use element checks if it should update its shadow-tree.

This now allows for the mutations in JavaScript to be visible!

Example Code:
```html
  <svg viewBox="0 0 30 10" xmlns="http://www.w3.org/2000/svg">
    <symbol id="myCircle">
    </symbol>
    <use href="#myCircle" fill="blue" />
    <use href="#myCircle" x="20" fill="white" stroke="red" />
  </svg>
  <svg id="delete"></svg>
  <script>
          let symbol = document.querySelector("symbol");
          function insert() {
                  let circle = document.createElementNS("http://www.w3.org/2000/svg", "circle");
                  circle.setAttribute("cx", "5");
                  circle.setAttribute("cy", "5");
                  circle.setAttribute("r", "4");
                  circle.setAttribute("stroke", "blue");
                  symbol.appendChild(circle);
          }
  
          function change_attribute() {
                  let circle = symbol.querySelector("circle");
                  circle.setAttribute("stroke", "red");
          }
  
          function remove() {
                  symbol.remove();
          }
  
          function remove_unrelated_svg() {
                  let svg = document.querySelector("#delete");
                  svg.remove();
          }
  
          setTimeout(insert, 2000);
          setTimeout(change_attribute, 4000);
          setTimeout(remove_unrelated_svg, 6000);
          setTimeout(remove, 8000);
  </script>
```